torch/distributed/_spmd/graph_optimization.py

```diff
@@ -548,3 +554,339 @@ def remove_copy_from_optimizer(gm: IterGraphModule) -> None:
+                    break
+            assert (
+                optim_getitem != self.optim_node
+            ), f"Cannot find the getitem node for {self.optim_node}"
+            output_list.extend(
+                [self.optim_node] * len(cast(List[fx.Node], self.optim_node.args[0]))
+            )
+            for updated_arg in optim_getitem.users:
+                assert (
+                    updated_arg.target == operator.getitem
+                ), f"Unexpected node target {updated_arg.target}."
+                idx = updated_arg.args[1]
+                output_copy = next(iter(updated_arg.users))
+                assert str(output_copy.target).startswith(
+                    "aten.copy_"
+                ), f"Unexpected node target {output_copy.target}."
+                output_list[idx] = output_copy
+            for i, output in enumerate(output_list):
+                assert output != self.optim_node, f"{i}th output is not replaced."
+
+            assert output_list, f"The output for {self.optim_node} is empty."
+
+        _populate_outputs(0, self.param_outputs)
+        _populate_outputs(2, self.exp_avgs_outputs)
+        _populate_outputs(3, self.exp_avg_sqs_outputs)
+
+    def __post_init__(self):
+        if self.param_outputs:
+            return
+        if self.generate_output:
+            self.generate_outputs()
+        else:
+            self.populate_outputs()
+
+
+@dataclass(unsafe_hash=True)
+class ForeachAddBlock:
+    add_node: fx.Node
+    generate_output: bool
+    # The output list of the copy nodes. The order follows the argument order.
+    outputs: List[fx.Node] = field(default_factory=list)
+
+    def generate_outputs(self):
+        # Iterate all the args and generate the corresponding output lists
+        # Assuming the corrsesponding output nodes are not created yet.
+        graph = self.add_node.graph
+        with graph.inserting_after(self.add_node):
+            for i, arg in enumerate(cast(Tuple[Any, ...], self.add_node.args[0])):
+                with graph.inserting_after(self.add_node):
+                    updated_arg = graph.call_function(
+                        operator.getitem, (self.add_node, i)
+                    )
+                with graph.inserting_after(updated_arg):
+                    output_copy = graph.call_function(aten.copy_, (arg, updated_arg))
+                self.outputs.append(output_copy)
+            assert self.outputs, f"The output for {self.add_node} is empty."
+
+    def populate_outputs(self):
+        # Populate the existing output lists from the graph.
+        self.outputs = [
+            self.add_node for _ in cast(Tuple[Any, ...], self.add_node.args[0])
+        ]
+        for updated_arg in self.add_node.users:
+            assert (
+                updated_arg.target == operator.getitem
+            ), f"Unexpected node target {updated_arg.target}"
+            idx = cast(int, updated_arg.args[1])
+            output_copy = next(iter(updated_arg.users))
```

Comment:
and this assumes the `copy_` is immediately after `getitem`? I suppose we fully control these order as we inserted those nodes and there won't be other passes be conducted in the middle of `generate_outputs` and `populate_outputs` to break these assumptions?

Reply:
If you mean topological order, it does not matter. We are using data dependency (users) to traverse.